pkg/reconciler/timeout_handler.go

```diff
@@ -101,6 +119,47 @@ func GetTimeout(d *metav1.Duration) time.Duration {
 	return timeout
 }
 
+// GetTaskRunEndTime calculates the time at which a TaskRun will be considered
+// timed out.
+func GetTaskRunEndTime(tr *v1alpha1.TaskRun) time.Time {
+	return tr.Status.StartTime.Time.Add(GetTimeout(tr.Spec.Timeout))
+}
+
+// GetBackoff records a count of the times it has seen a given StatusKey object
+// and calculates an appropriate backoff deadline based on that count. Only one
+// deadline per StatusKey may be active at any moment. Requests for a new backoff
+// in the face of an existing one will be ignored and details of the existing
+// backoff will be returned instead.
+//
+// Returned values are: the count of backoffs performed for this StatusKey, the
+// backoff's deadline and a boolean indicating whether the returned backoff is
+// newly calculated.
+func (t *TimeoutSet) GetBackoff(runObj StatusKey) (uint, time.Time, bool) {
+	t.backoffsMut.Lock()
+	defer t.backoffsMut.Unlock()
+	b := t.backoffs[runObj.GetRunKey()]
+	if time.Now().Before(b.deadline) {
+		return b.count, b.deadline, false
+	}
+	b.count += 1
+	b.deadline = time.Now().Add(backoffDuration(b.count))
+	t.backoffs[runObj.GetRunKey()] = b
+	return b.count, b.deadline, true
```

Comment:
The `backoff` struct felt to me like an implementation detail of the internal state of the backoff-tracking mechanism and I didn't sense a strong reason to expose a new public type in the package's interface.  I don't feel strongly about this and am happy to change it but am curious what might be the benefit of making it public?

Reply:
> but am curious what might be the benefit of making it public?

The only one I can think of is simplifying the signature of the function a bit - since you already have a structure!

It reminded me of a pattern I've often seen where a function takes too many parameters or returns too many things, and this is often a sign that another abstraction can be added that would simplify the interface (this website talks about it a bit: https://refactoring.guru/smells/long-parameter-list disclaimer I have never seen this website before haha) - and it seemed like what was happening here was doing the opposite

Just a super minor thing tho!